be/src/storage/lake/txn_log_applier.cpp

```diff
@@ -0,0 +1,270 @@
+        if (op_compaction.has_output_rowset() && op_compaction.output_rowset().num_rows() > 0) {
+            // Replace the first input rowset with output rowset
+            auto output_rowset = _metadata->mutable_rowsets(first_idx);
+            output_rowset->CopyFrom(op_compaction.output_rowset());
+            output_rowset->set_id(_metadata->next_rowset_id());
+            _metadata->set_next_rowset_id(_metadata->next_rowset_id() + output_rowset->segments_size());
+            ++first_input_pos;
+        }
+        // Erase input rowsets from _metadata
+        auto end_input_pos = pre_input_pos + 1;
+        _metadata->mutable_rowsets()->erase(first_input_pos, end_input_pos);
+
+        // Set new cumulative point
+        uint32_t new_cumulative_point = 0;
+        if (first_idx >= _metadata->cumulative_point()) {
+            // cumulative compaction
+            new_cumulative_point = first_idx;
+        } else {
+            // base compaction
+            new_cumulative_point = _metadata->cumulative_point() - op_compaction.input_rowsets_size();
+        }
+        if (op_compaction.has_output_rowset() && op_compaction.output_rowset().num_rows() > 0) {
+            ++new_cumulative_point;
+        }
+        if (new_cumulative_point > _metadata->rowsets_size()) {
+            return Status::InternalError(fmt::format("new cumulative point: {} exceeds rowset size: {}",
+                                                     new_cumulative_point, _metadata->rowsets_size()));
+        }
+        _metadata->set_cumulative_point(new_cumulative_point);
+
+        // Debug new tablet _metadata
+        std::vector<uint32_t> rowset_ids;
+        std::vector<uint32_t> delete_rowset_ids;
+        for (const auto& rowset : _metadata->rowsets()) {
+            rowset_ids.emplace_back(rowset.id());
+            if (rowset.has_delete_predicate()) {
+                delete_rowset_ids.emplace_back(rowset.id());
+            }
+        }
+        LOG(INFO) << "compaction finish. tablet: " << _metadata->id() << ", version: " << _metadata->version()
+                  << ", cumulative point: " << _metadata->cumulative_point() << ", rowsets: ["
+                  << JoinInts(rowset_ids, ",") << "]"
+                  << ", delete rowsets: [" << JoinInts(delete_rowset_ids, ",") + "]";
+        return Status::OK();
+    }
+
+    Status apply_schema_change_log(const TxnLogPB_OpSchemaChange& op_schema_change) {
+        for (const auto& rowset : op_schema_change.rowsets()) {
+            auto new_rowset = _metadata->add_rowsets();
+            new_rowset->CopyFrom(rowset);
+            new_rowset->set_id(_metadata->next_rowset_id());
+            _metadata->set_next_rowset_id(_metadata->next_rowset_id() + std::max(1, new_rowset->segments_size()));
+        }
+        if (op_schema_change.has_delvec_meta()) {
+            _metadata->mutable_delvec_meta()->CopyFrom(op_schema_change.delvec_meta());
+        }
```

Comment:
non primary key table does not have delvec meta

Reply:
Updated, PTAL